File: java/IsValidBST.java
// https://leetcode.com/problems/validate-binary-search-tree/

import container.TreeNode;

public class IsValidBST {
    /**
     * Check if a binary tree is a valid binary search tree (BST).
     *
     * <p>We process the answer recursively with in-order traversal. The idea is to
     * look for a previous entry if it exists and see if that entry is less than,
     * or equal to, the current entry.
     *
     * <p>Complexity:
     *
     * <ul>
     *     <li>Time: O(n)</li>
     *     <li>Space: O(h)</li>
     * </ul>
     */
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, null, null);
    }

    /**
     * Helper method that validates BST property with min and max bounds.
     * @param node Current node being validated
     * @param min Minimum allowed value (exclusive)
     * @param max Maximum allowed value (exclusive)
     */
    private boolean isValidBST(TreeNode node, Integer min, Integer max) {
        if (node == null) {
            return true;
        }

        // Check if current node violates BST property
        if ((min != null && node.val <= min) || (max != null && node.val >= max)) {
            return false;
        }

        // Recursively validate left and right subtrees with updated bounds
        return isValidBST(node.left, min, node.val)
            && isValidBST(node.right, node.val, max);
    }
}

File: python/isValidBST.py
# https://leetcode.com/problems/validate-binary-search-tree/

from typing import Optional

from model.binary_tree import TreeNode


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        """
        Check if the binary tree is a valid BST.

        We use an iterative in-order traversal to check if the values are in strictly
        increasing order.

        Complexity:
        - Time: O(n)
        - Space: O(n)
        """
        st: list[TreeNode] = []
        curr = root
        prev = None

        while len(st) > 0 or curr:
            while curr:
                st.append(curr)
                curr = curr.left

            # Pop once to emulate an in-order visit
            curr = st.pop()

            if prev and prev.val >= curr.val:
                return False

            # Look at the right subtree of each node
            prev = curr
            curr = curr.right

        return True
